### ISSUE TITLE
NFS - SHIFT - running 2x as fast in Ad-hoc

### ISSUE THREAD

**Issue by samosaphile** [hrydgard/ppsspp#14661]
### What games and where in them does it happen?

ULUS10462.

### What speed are you getting with PPSSPP speed display enabled?
Well, it shows the same FPS as in offline, but clearly something is up in ad-hoc netplay mode cuz it's uncontrollable compared to single player.

#### What hardware / device and operating system?
Pixel 4a and stock android 11.
#### What graphics card (GPU) or mobile phone model?
Adreno 618.
#### What PPSSPP version (standalone/official), and did it work before?
I'm not aware it ever worked correctly. Tested on both 1.11.2-432 and 1.11.3-842. But again, pretty sure this doesn't matter - I've tested this multiple times with multiple people (iOS, android, pc, both US and EU players) doesn't seem to matter. Seems network related as game behaves completely different in ad-hoc mode.
#### Graphics backend (Vulkan, OpenGL, etc.)?
Don't think this is a backend issue as works fine in offline modes but tested on vulkan.

#### Checklist

 - [ ] Test in the [latest git build](https://buildbot.orphis.net/ppsspp/) in case it's already fixed.
 - [ x] [Search for other reports](https://github.com/hrydgard/ppsspp/search?q=my+issue&type=issues) of the same issue.
 - [x ] Try resetting settings or older versions and include if the issue is related.
 - [ ] Try changing graphics settings to determine if one causes the slowdown.
 - [ ] Include logs of screenshots of issue.

**Comment by ANR2ME** [comment]
Probably similar issue to Bleach 7, Fate Unlimited Codes, and Burnout, where FPS is running at 60 FPS instead of 30 FPS during multiplayer.

We haven't figured out why it's happening since i didn't have the time yet to investigate the game code that calculate the Vblank number.

**Comment by MojoJojoDojo** [comment]
At least on Burnout this also happens on real hardware when you are trying to play online (Adhoc Party) so its not entirely seems like a bug...
Then again there it probably works fine locally.
Changing the Emulated CPU clock speed in System to below 100 mhz seems to temporarily fix it though at least on stable 1.11.3.

**Comment by unknownbrackets** [comment]
If this happens on real hardware, it's not a PPSSPP bug but a game bug, and should be patched in the game.

-[Unknown]

**Comment by MojoJojoDojo** [comment]
![image](https://user-images.githubusercontent.com/33392696/127758284-96f4fe98-834f-40d1-adf6-3221ff933f8c.png)

This is only true for Dominator for now.
Someone will need to try the games on localhost to see if its not an online only issue though.

**Comment by ANR2ME** [comment]
> 
> 
> ![image](https://user-images.githubusercontent.com/33392696/127758284-96f4fe98-834f-40d1-adf6-3221ff933f8c.png)
> 
> This is only true for Dominator for now but could apply to Shift as well (Both made by EA UK team).
> Someone will need to try the games on localhost to see if its not an online only issue though.
> It could still happen on localhost though because PPSSPP does not truly emulates a locally connected game.
> If someone doesnt know , Ad-hoc Party is an PS3 app that tunnels PSP's ad-hoc to work over the internet.
> BTW there is at least 1 game that doesn't work with it but does work on PPSSPP lol.
> 
> In general EA games seems to have buggy Ad-hoc mode.
> If you try to play with more then 3 players on ad-hoc on both xlink and Ad-hoc Party with Burnout Legends the game might crash randomly lol.

This guy seems to have similar issue with Burnout Dominator on PSP with CFW https://community.eurogamer.net/thread/122528
While his friend (which use the same CFW, and the same ISO based on his later post) seems to work fine, so may be it have something to do with his settings (ie. custom CPU clock, etc.) causing this speedup...

**Comment by MojoJojoDojo** [comment]
> > ![image](https://user-images.githubusercontent.com/33392696/127758284-96f4fe98-834f-40d1-adf6-3221ff933f8c.png)
> > This is only true for Dominator for now but could apply to Shift as well (Both made by EA UK team).
> > Someone will need to try the games on localhost to see if its not an online only issue though.
> > It could still happen on localhost though because PPSSPP does not truly emulates a locally connected game.
> > If someone doesnt know , Ad-hoc Party is an PS3 app that tunnels PSP's ad-hoc to work over the internet.
> > BTW there is at least 1 game that doesn't work with it but does work on PPSSPP lol.
> > In general EA games seems to have buggy Ad-hoc mode.
> > If you try to play with more then 3 players on ad-hoc on both xlink and Ad-hoc Party with Burnout Legends the game might crash randomly lol.
> 
> This guy seems to have similar issue with Burnout Dominator on PSP with CFW https://community.eurogamer.net/thread/122528
> While on his friend's PSP (which use the same CFW, and the same ISO based on his later post) seems to work fine, so may be it have something to do with his settings (ie. custom CPU clock, etc.) causing this speedup...
> 
> PS: They doesn't seems to be talking about Adhoc Party

Hmm could be but you glossed over the fact he only got that issue on multiplayer....
Also unlike the other games like Bleach etc the FPS counter on PPSSPP stays the same between offline and online (like the issue opener said).
The one who opened the issue didn't say it affects all the players or only the host player so it could still be related somehow...
It does lends to some sort of game bug that the game got in a way.

**Comment by MojoJojoDojo** [comment]
Just tried playing with someone over xlink kai to confirm what happens on hardware myself.
The game was too fast for me and the other player (with some minor drops).
I was using PSP Go and the other user used a VIta.
At least on my end the CPU clocks were on Default but not sure about the other player.

**Comment by ANR2ME** [comment]
> 
> 
> Just tried playing with someone over xlink kai to confirm what happens on hardware myself.
> The game was too fast for me and the other player (it seems even way more faster for the other player though).
> I was using PSP Go and the other user used a VIta.
> At least on my end the CPU clocks were on Default but not sure about the other player.
> Weird stuff indeed.

Which game did you test? NFS or Burnout?

**Comment by MojoJojoDojo** [comment]
> > Just tried playing with someone over xlink kai to confirm what happens on hardware myself.
> > The game was too fast for me and the other player (it seems even way more faster for the other player though).
> > I was using PSP Go and the other user used a VIta.
> > At least on my end the CPU clocks were on Default but not sure about the other player.
> > Weird stuff indeed.
> 
> Which game did you test? NFS or Burnout?

Burnout but again they are both done by the exact same developers.
They only developed these 2 games on PSP as well if you dont count the HP games.
High chance that they use either the same netcode or engine.

**Comment by samosaphile** [comment]
Yeah, I think dominator ran too fast for me too. Not legends tho, that seemed fine.

**Comment by MojoJojoDojo** [comment]
Tried it again PSP vs PSP on xlink and couldn't trigger the speedup behavior.
Because I did tried it with a Vita user it does seem to be tied to CPU clocks somehow I guess.

**Comment by ANR2ME** [comment]
May be the game is running at a different CPU clock when in Vita? 
i wonder if each players are choosing a different PSP Model on PPSSPP settings will also affects the CPU clock...

**Comment by MojoJojoDojo** [comment]
It won't but setting the emulated CPU clock on System settings at around 50 mhz or less on one or both instances will make the game go down to around 30 fps somewhat.
It's easier to do it when the HLE delay was larger so on the latest stable build you can prolly up the clocks to 100 mhz or more.

**Comment by metehan989** [comment]
Guys the glitches and bugs only occur on ppsspp. This isnt a hardware issue on the realsystems. All my psp games run smoothly on my psp and also on my ps vita. They also dont have any connection issues or multiplayer issues. I think its really ppsspp having a bug. I last checked this today. The only game i do know that bugs on the psp is spiderman 3. The audio disappears on a certain part of the story mode. You wont receive any audio after. I havent tested this on ppsspp tho so i would appreciate someone testing it for us. Thats all i know for now.

-metehan989

**Comment by MojoJojoDojo** [comment]
> Guys the glitches and bugs only occur on ppsspp. This isnt a hardware issue on the realsystems. All my psp games run smoothly on my psp and also on my ps vita. They also dont have any connection issues or multiplayer issues. I think its really ppsspp having a bug. I last checked this today. The only game i do know that bugs on the psp is spiderman 3. The audio disappears on a certain part of the story mode. You wont receive any audio after. I havent tested this on ppsspp tho so i would appreciate someone testing it for us. Thats all i know for now.
> 
> -metehan989

Me and @ANR2ME already showed proof that it does happen on console in some cases but that PPSSPP keep triggering thar behaviour all the time.
I also didn't say its a hardware issue but rather a game issue.

**Comment by TurkishG** [comment]
it doesnt it also occurs on real hardware

**Comment by hrydgard** [comment]
I think we can all agree that it seems to occur a lot more on PPSSPP than on hardware, at least. The reason for that might be a bug with some kind of synchronization, might be differences in speed of emulation vs real hardware, or any number of things.. But it's an issue.

**Comment by MojoJojoDojo** [comment]
BTW noticed another weird thing recently with this game..
Usually when someone disconnects the game will disconnect for all other players.
But it won't disconnect while playing it.
Also the game starts weirdly- there is no countdown until the race starts and the cars are placed randomly on the track...

**Comment by MojoJojoDojo** [comment]
Well the owner of the issue never posted how it looks ingame so here is footage from hardware that gets that bug for Burnout at least:
PSP side : https://www.youtube.com/watch?v=Fbeaqwpj-lU
Vita side:https://youtu.be/9LQJaB8dtY8
When the Vita user or users stopped playing or when they downclock the CPU clocks from default the bug won't occur.
The reason the other Vita user in this issue didn't get it is might be because he didn't Adrenaline (the others in the video did).
